pkg/handlers/generic/lifecycle/cni/cilium/handler.go

```diff
@@ -259,65 +261,145 @@ func runApply(
 		return err
 	}
 
+	// It is possible to disable kube-proxy and migrate to Cilium's kube-proxy replacement feature in a running cluster.
+	// In this case, we need to wait for Cilium to be restarted with new configuration and then cleanup kube-proxy.
+
 	// If skip kube-proxy is not set, return early.
-	// Otherwise, wait for Cilium to be rolled out and then cleanup kube-proxy if installed.
 	if !capiutils.ShouldSkipKubeProxy(cluster) {
 		return nil
 	}
 
+	remoteClient, err := remote.NewClusterClient(
+		ctx,
+		"",
+		client,
+		ctrlclient.ObjectKeyFromObject(cluster),
+	)
+	if err != nil {
+		return fmt.Errorf("error creating remote cluster client: %w", err)
+	}
+
+	// If kube-proxy is not installed,
+	// assume that the one-time migration of kube-proxy is complete and return early.
+	isKubeProxyInstalled, err := isKubeProxyInstalled(ctx, remoteClient)
```

Comment:
More of a note to self: When `cleanupKubeProxy` returns, all we know is that the API server has set the deletion timestamp on the resources. They may still be around the next time the handler is called. Then, `isKubeProxyInstalled` will return true, and we'll end up running the remainder of the code, including `cleanupKubeProxy`, again. But since `forceCiliumRollout` only sets the "restartedAt" annotation once,  this is safe.